resource/resource.go

```diff
@@ -168,46 +151,145 @@ func (r *Resource) Out(inputDirectory string, source oc.Source, params oc.Params
 
-	err = filepath.Walk(rootDir,
-		func(path string, info os.FileInfo, err error) error {
-			if info.IsDir() {
-				return nil
+	for _, secretMap := range p.SecretMaps {
+
+		sourceDir := filepath.Join(rootDir, secretMap.Source)
+		files, err := listFilesUnder(sourceDir)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		for _, secretFile := range files {
+			secret, err := createSecret(sourceDir, secretFile)
+			if err != nil {
+				return nil, nil, err
 			}
 
-			bytes, err := ioutil.ReadFile(path)
+			err = validate(secret, secretMap.Keys)
 			if err != nil {
-				return err
+				return nil, nil, err
+			}
+
+			err = filterAndRenameKeys(secret, secretMap.Keys)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			err = copySecretToVault(r.client, p.Prefix, secretMap.Dest, secretFile, secret)
+			if err != nil {
+				return nil, nil, err
 			}
-			secert := sv.NewSecret()
-			err = secert.UnmarshalJSON(bytes)
+
+		}
+	}
+	// Both `version` and `metadata` may be empty. In this case, we are returning
+	// `version` retrieved from the file created by `In`, while `metadata` is empty.
+	metadata := oc.Metadata{}
+	return nil, metadata, nil
+}
+
+func listFilesUnder(rootDir string) ([]string, error) {
+	ret := []string{}
+	err := filepath.Walk(rootDir,
+		func(path string, info os.FileInfo, err error) error {
 			if err != nil {
 				return err
 			}
+			if info.IsDir() {
+				return nil
+			}
 			secretPath, err := filepath.Rel(rootDir, path)
 			if err != nil {
 				return err
 			}
-			return r.client.Write(filepath.Join(p.Prefix, secretPath), secert)
-		})
+			ret = append(ret, secretPath)
+			return nil
+		},
+	)
+	return ret, err
+}
+
+func createSecret(rootDir string, secretFile string) (*sv.Secret, error) {
+	srcFile, err := os.Open(filepath.Join(rootDir, secretFile))
 	if err != nil {
-		return nil, nil, err
+		return nil, fmt.Errorf("Error reading source file '%s': %s", secretFile, err)
+	}
+	jDec := json.NewDecoder(srcFile)
+	secret := sv.NewSecret()
+	err = jDec.Decode(&secret)
+	if err != nil {
+		return nil, err
 	}
+	return secret, nil
+}
 
-	// Both `version` and `metadata` may be empty. In this case, we are returning
-	// `version` retrieved from the file created by `In`, while `metadata` is empty.
-	metadata := oc.Metadata{}
-	return nil, metadata, nil
+func validate(secret *sv.Secret, keys []interface{}) error {
+	finalKeys := getFinalKeys(keys)
+	ok := true
+	message := "Specified keys not found:"
+	for key := range finalKeys {
+		if !secret.Has(key) {
+			ok = false
+			message += fmt.Sprintf(" '%s'", key)
+		}
+	}
+	if !ok {
+		return fmt.Errorf(message)
+	}
+	return nil
+}
+
+func filterAndRenameKeys(secret *sv.Secret, keys []interface{}) error {
+	if len(keys) == 0 {
+		return nil
+	}
+	finalKeys := getFinalKeys(keys)
+	for _, currentKey := range secret.Keys() {
+		finalKey, isFinalKey := finalKeys[currentKey]
+		if isFinalKey {
+			value := secret.Get(currentKey)
+			err := secret.Set(finalKey, value, false)
```

Comment:
how does this handle the case where the user wants to rename, for example
foo->bar
bar->beep
beep->foo
?

The way I see it, if a user tries to rename a key to the name of an existing key, there's a race condition (because looping over a map in go is non-deterministic in ordering) as to what gets overwritten.

Reply:
this still needs to be addressed if we're not making reusing key names illegal.